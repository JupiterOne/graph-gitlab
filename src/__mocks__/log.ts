export const debug = jest.fn();
export const info = jest.fn();
export const warn = jest.fn();
export const displayExecutionResults = jest.fn();
