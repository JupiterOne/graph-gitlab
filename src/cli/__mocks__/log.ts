import noop from 'lodash/noop';

export const debug = noop;
export const info = noop;
