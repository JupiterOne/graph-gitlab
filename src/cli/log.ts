import chalk from 'chalk';

export function debug(msg: string) {
  console.log(`${chalk.gray(msg)}`);
}

export function info(msg: string) {
  console.log(`${chalk.cyan(msg)}`);
}
