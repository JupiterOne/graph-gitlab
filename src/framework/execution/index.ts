export * from './executeIntegration';
export * from './types';
