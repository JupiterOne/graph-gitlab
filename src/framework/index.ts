export * from './data';
export * from './execution';
export * from './visualization';
export * from './types';
