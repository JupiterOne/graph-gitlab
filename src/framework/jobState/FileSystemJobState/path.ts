export type CollectionType = 'entities' | 'relationships';

interface BuildStepCollectionFilePathInput {
  step: string;
  collectionType: CollectionType;
  filename: string;
}

export function buildStepCollectionFilePath({
  step,
  collectionType,
  filename,
}: BuildStepCollectionFilePathInput) {
  return ['graph', step, collectionType, filename].join('/');
}

interface BuildIndexFilePathInput extends BuildIndexDirectoryPathInput {
  filename: string;
}

export function buildIndexFilePath({
  collectionType,
  type,
  filename,
}: BuildIndexFilePathInput) {
  return [buildIndexDirectoryPath({ collectionType, type }), filename].join(
    '/',
  );
}

interface BuildIndexDirectoryPathInput {
  collectionType: CollectionType;
  type: string;
}

export function buildIndexDirectoryPath({
  collectionType,
  type,
}: BuildIndexDirectoryPathInput) {
  return ['index', collectionType, type].join('/');
}
