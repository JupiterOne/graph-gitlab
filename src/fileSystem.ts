/**
 * Home of all file system interactions
 *
 * This module exports utilities for writing data
 * relative to the .j1-integration root storage directoryPath.
 */
import * as nodeFs from 'fs';
const fs = nodeFs.promises;
import path from 'path';

import rimraf from 'rimraf';

export const DEFAULT_CACHE_DIRECTORY_NAME = '.j1-integration';

export function getRootStorageDirectory() {
  return path.resolve(process.cwd(), DEFAULT_CACHE_DIRECTORY_NAME);
}

interface WriteDataToPathInput {
  path: string;
  data: object;
}

/**
 * Function for writing arbirary data to a path
 * relative to the cache directory.
 *
 * This will ensure that the directories exists or have been
 * created prior to writing the file.
 */
export async function writeJsonToPath({
  path: relativePath,
  data,
}: WriteDataToPathInput) {
  const directory = getRootStorageDirectory();
  const fullPath = path.resolve(directory, relativePath);

  await ensurePathCanBeWrittenTo(fullPath);
  await fs.writeFile(fullPath, JSON.stringify(data, null, 2), 'utf8');
}

interface SymlinkInput {
  sourcePath: string;
  destinationPath: string;
}

/**
 * Function for creating a symlink from on file to another.
 *
 * This will ensure that the directories exists or have been
 * created prior to writing the file.
 */
export async function symlink({ sourcePath, destinationPath }: SymlinkInput) {
  const directory = getRootStorageDirectory();
  const fullSourcePath = path.resolve(directory, sourcePath);
  const fullDestinationPath = path.resolve(directory, destinationPath);

  await ensurePathCanBeWrittenTo(fullDestinationPath);
  await fs.symlink(fullSourcePath, fullDestinationPath);
}

export interface WalkDirectoryIterateeInput {
  filePath: string;
  data: string;
}

type WalkDirectoryIteratee = (
  input: WalkDirectoryIterateeInput,
) => Promise<void> | void;

interface WalkDirectoryInput {
  path: string;
  iteratee: WalkDirectoryIteratee;
}

/**
 * Function for recursively walking through a directory
 * and reading the data from each file.
 */
export async function walkDirectory({
  path: relativePath,
  iteratee,
}: WalkDirectoryInput) {
  const directory = getRootStorageDirectory();
  const fullPath = path.resolve(directory, relativePath);

  const isDirectory = await isDirectoryPresent(fullPath);
  if (!isDirectory) {
    return;
  }

  const files = await fs.readdir(fullPath);

  const onFile = async (filePath: string) => {
    const data = await fs.readFile(filePath, 'utf8');
    await iteratee({ filePath, data });
  };

  const handleFilePath = async (filePath: string) => {
    const stats = await fs.lstat(filePath);
    if (stats.isDirectory()) {
      // continue walking the directory
      await walkDirectory({
        iteratee,
        path: filePath,
      });
    } else if (stats.isFile()) {
      // handle the file
      await onFile(filePath);
    } else if (stats.isSymbolicLink()) {
      // resolve the symlink then reperform check
      // to determine path resolves to a file or
      // if we should continue recursing
      const realPath = await fs.realpath(filePath);
      await handleFilePath(realPath);
    }
  };

  for (const file of files) {
    await handleFilePath(path.resolve(fullPath, file));
  }
}

/**
 * Wipes the storage directory clean
 */
export async function removeStorageDirectory() {
  const rootStorageDir = getRootStorageDirectory();
  if (await isDirectoryPresent(rootStorageDir)) {
    await removeDirectory(rootStorageDir);
  }
}

function removeDirectory(directory: string) {
  return new Promise((resolve, reject) =>
    rimraf(directory, (err) => {
      if (err) {
        return reject(err);
      }
      resolve();
    }),
  );
}

async function isDirectoryPresent(fullPath: string) {
  try {
    const stats = await fs.lstat(fullPath);
    return stats.isDirectory();
  } catch (err) {
    if (err.code === 'ENOENT') {
      return false;
    }

    // not what we expected... throw error
    throw err;
  }
}

async function ensurePathCanBeWrittenTo(pathToWrite: string) {
  const directoryPath = path.dirname(pathToWrite);
  await fs.mkdir(directoryPath, { recursive: true });
}
