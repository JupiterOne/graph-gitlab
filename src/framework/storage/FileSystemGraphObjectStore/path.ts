import path from 'path';

export type CollectionType = 'entities' | 'relationships';

interface BuildObjectCollectionFilePathInput {
  storageDirectoryPath: string;
  collectionType: CollectionType;
  filename: string;
}

export function buildObjectCollectionFilePath({
  storageDirectoryPath,
  collectionType,
  filename,
}: BuildObjectCollectionFilePathInput) {
  return path.join('graph', storageDirectoryPath, collectionType, filename);
}

interface BuildIndexFilePathInput extends BuildIndexDirectoryPathInput {
  filename: string;
}

export function buildIndexFilePath({
  collectionType,
  type,
  filename,
}: BuildIndexFilePathInput) {
  return path.join(buildIndexDirectoryPath({ collectionType, type }), filename);
}

interface BuildIndexDirectoryPathInput {
  collectionType: CollectionType;
  type: string;
}

export function buildIndexDirectoryPath({
  collectionType,
  type,
}: BuildIndexDirectoryPathInput) {
  return path.join('index', collectionType, type);
}
