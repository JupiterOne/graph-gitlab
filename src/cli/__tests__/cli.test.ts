import path from 'path';
import { createCli } from '../index';
import { loadProjectStructure } from '../../__tests__/loadProjectStructure';
import * as log from '../../log';

import { IntegrationStepResultStatus } from '../../framework/execution';
import { getRootStorageDirectory } from '../../fileSystem';
import * as nodeFs from 'fs';
const fs = nodeFs.promises;

jest.mock('../../log');

afterEach(() => {
  // clear require cache
  Object.keys(require.cache).forEach((modulePath) => {
    delete require.cache[modulePath];
  });

  delete process.env.MY_CONFIG;
});

describe('collect', () => {
  beforeEach(() => {
    loadProjectStructure('typeScriptProject');
  });

  test('loads the integration, executes it, and logs the result', async () => {
    await createCli().parseAsync(['node', 'j1-integration', 'collect']);

    expect(log.displayExecutionResults).toHaveBeenCalledTimes(1);
    expect(log.displayExecutionResults).toHaveBeenCalledWith({
      integrationStepResults: [
        {
          id: 'fetch-accounts',
          name: 'Fetch Accounts',
          types: ['my_account'],
          status: IntegrationStepResultStatus.SUCCESS,
        },
        {
          id: 'fetch-users',
          name: 'Fetch Users',
          types: ['my_user'],
          status: IntegrationStepResultStatus.SUCCESS,
        },
      ],
      metadata: {
        partialDatasets: {
          types: [],
        },
      },
    });
  });
});

describe('visualize', () => {
  beforeEach(() => {
    loadProjectStructure('typeScriptVisualizeProject');
  });

  test('writes graph to html file', async () => {
    await createCli().parseAsync(['node', 'j1-integration', 'visualize']);

    const htmlFileLocation = path.join(getRootStorageDirectory(), 'index.html');
    const content = await fs.readFile(htmlFileLocation, 'utf8');

    expect(log.info).toHaveBeenCalledWith(
      `Visualize Integration SDK graph here: ${htmlFileLocation}`,
    );
    expect(content).toMatchSnapshot();
  });
});

describe('collect/visualize integration', () => {
  beforeEach(() => {
    loadProjectStructure('typeScriptIntegrationProject');
  });

  test('creates graph based on integration data', async () => {
    await createCli().parseAsync(['node', 'j1-integration', 'collect']);
    await createCli().parseAsync(['node', 'j1-integration', 'visualize']);

    const htmlFileLocation = path.join(getRootStorageDirectory(), 'index.html');
    const content = await fs.readFile(htmlFileLocation, 'utf8');

    expect(log.info).toHaveBeenCalledWith(
      `Visualize Integration SDK graph here: ${htmlFileLocation}`,
    );
    expect(content).toMatchSnapshot();
  });
});
