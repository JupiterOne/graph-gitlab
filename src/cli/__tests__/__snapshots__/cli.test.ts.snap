// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`collect/visualize integration creates graph based on integration data 1`] = `
"<!DOCTYPE html>
    <html lang=\\"en\\">
    <head>
        <meta charset=\\"UTF-8\\">
        <meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1.0\\">
        <title>Integration SDK Graph</title>
    </head>
    <body style=\\"height:100vh;width:100%\\">
        <div id=\\"integration-graph\\" style=\\"height:100vh;width:100%\\"></div>
        <script type=\\"text/javascript\\" src=\\"https://unpkg.com/vis-network/standalone/umd/vis-network.min.js\\"></script>
        <script type=\\"text/javascript\\">
            var nodes = new vis.DataSet([{\\"id\\":\\"account:1234\\",\\"label\\":\\"Account 1234\\"},{\\"id\\":\\"user:12345\\",\\"label\\":\\"User 12345\\"}]);
            var edges = new vis.DataSet([{\\"from\\":\\"account:1234\\",\\"to\\":\\"user:12345\\",\\"label\\":\\"HAS\\"}]);
            // create a network
            var container = document.getElementById(\\"integration-graph\\");
            var data = {
              nodes: nodes,
              edges: edges
            };
            var options = {\\"edges\\":{\\"arrows\\":{\\"to\\":{\\"enabled\\":true}}}};
            var network = new vis.Network(container, data, options);
          </script>
    </body>
    </html>"
`;

exports[`visualize writes graph to html file 1`] = `
"<!DOCTYPE html>
    <html lang=\\"en\\">
    <head>
        <meta charset=\\"UTF-8\\">
        <meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1.0\\">
        <title>Integration SDK Graph</title>
    </head>
    <body style=\\"height:100vh;width:100%\\">
        <div id=\\"integration-graph\\" style=\\"height:100vh;width:100%\\"></div>
        <script type=\\"text/javascript\\" src=\\"https://unpkg.com/vis-network/standalone/umd/vis-network.min.js\\"></script>
        <script type=\\"text/javascript\\">
            var nodes = new vis.DataSet([{\\"id\\":\\"entity:1\\",\\"label\\":\\"Entity Name\\"},{\\"id\\":\\"entity:2\\",\\"label\\":\\"Entity Name\\"}]);
            var edges = new vis.DataSet([{\\"from\\":\\"entity:1\\",\\"to\\":\\"entity:2\\",\\"label\\":\\"HAS\\"}]);
            // create a network
            var container = document.getElementById(\\"integration-graph\\");
            var data = {
              nodes: nodes,
              edges: edges
            };
            var options = {\\"edges\\":{\\"arrows\\":{\\"to\\":{\\"enabled\\":true}}}};
            var network = new vis.Network(container, data, options);
          </script>
    </body>
    </html>"
`;
